usage/src/org/apache/cloudstack/quota/QuotaAlertManagerImpl.java

```diff
@@ -0,0 +1,444 @@
+                }
+            }
+        } else {
+            s_logger.warn("Failed to lock account " + accountId + ", account not found.");
+        }
+        TransactionLegacy.open(opendb).close();
+        return success;
+    }
+
+    public static class DeferredQuotaEmail {
+        private AccountVO account;
+        private QuotaAccountVO quotaAccount;
+        private QuotaConfig.QuotaEmailTemplateTypes emailTemplateType;
+        private BigDecimal quotaUsage;
+
+        public DeferredQuotaEmail(AccountVO account, QuotaAccountVO quotaAccount, BigDecimal quotaUsage, QuotaConfig.QuotaEmailTemplateTypes emailTemplateType) {
+            this.account = account;
+            this.quotaAccount = quotaAccount;
+            this.emailTemplateType = emailTemplateType;
+            this.quotaUsage = quotaUsage;
+        }
+
+        public DeferredQuotaEmail(AccountVO account, QuotaAccountVO quotaAccount, QuotaConfig.QuotaEmailTemplateTypes emailTemplateType) {
+            this.account = account;
+            this.quotaAccount = quotaAccount;
+            this.emailTemplateType = emailTemplateType;
+            this.quotaUsage = new BigDecimal(-1);
+        }
+
+        public AccountVO getAccount() {
+            return account;
+        }
+
+        public BigDecimal getQuotaBalance() {
+            return quotaAccount.getQuotaBalance();
+        }
+
+        public BigDecimal getQuotaUsage() {
+            return quotaUsage;
+        }
+
+        public QuotaConfig.QuotaEmailTemplateTypes getEmailTemplateType() {
+            return emailTemplateType;
+        }
+
+        public void sentSuccessfully(final QuotaAccountDao quotaAccountDao) {
+            if (emailTemplateType == QuotaEmailTemplateTypes.QUOTA_STATEMENT) {
+                quotaAccount.setLastStatementDate(new Date());
+            } else {
+                quotaAccount.setQuotaAlertDate(new Date());
+                quotaAccount.setQuotaAlertType(emailTemplateType.ordinal());
+            }
+            quotaAccountDao.update(quotaAccount.getAccountId(), quotaAccount);
+        }
+    };
+
+    class EmailQuotaAlert {
+        private Session _smtpSession;
+        private final String _smtpHost;
+        private final int _smtpPort;
+        private final boolean _smtpUseAuth;
+        private final String _smtpUsername;
+        private final String _smtpPassword;
+        private final String _emailSender;
+
+        public EmailQuotaAlert(String smtpHost, int smtpPort, boolean smtpUseAuth, final String smtpUsername, final String smtpPassword, String emailSender, boolean smtpDebug) {
+            _smtpHost = smtpHost;
+            _smtpPort = smtpPort;
+            _smtpUseAuth = smtpUseAuth;
+            _smtpUsername = smtpUsername;
+            _smtpPassword = smtpPassword;
+            _emailSender = emailSender;
+
+            if (_smtpHost != null) {
```

Comment:
What is _smtpHost is blank?  Check with ``com.google.common.base.Strings.isNullorEmpty()`` to protect against this condition.